java/src/test/java/com/google/appengine/tools/pipeline/impl/servlets/JsonGeneratorTest.java

```diff
@@ -0,0 +1,83 @@
+
+  private static class ConcreteJob extends Job0 {
+
+    @Override
+    public String getJobDisplayName() {
+      return "ConcreteJob";
+    }
+
+    @SuppressWarnings("unused")
+    public Value<String> handleException(Throwable t) {
+      return immediate("Got exception!");
+    }
+
+    @Override
+    public Value run() throws Exception {
+      return immediate(null);
+    }
+  }
+
+  //TODO: this stuff should all be mocked, so can control jobIds + timestamps, more properly validate JSON
+  // outputs
+  PipelineObjects exampleObjects() throws Exception {
+
+    PipelineObjects exampleObjects = mock(PipelineObjects.class);
+
+    PipelineService service = PipelineServiceFactory.newPipelineService();
+    ConcreteJob job = new ConcreteJob();
+
+    //job ids seem to auto-inc in stubbed local data store, so beware ...
+    String pipelineId = service.startNewPipeline(job);
+    JobRecord jobRecord = PipelineManager.getJob(pipelineId);
+    JobInfo jobInfo = waitUntilJobComplete(pipelineId);
+    jobRecord = PipelineManager.getJob(pipelineId);
+    return PipelineManager.queryFullPipeline(pipelineId);
+  }
+
+  public void testMap() throws Exception {
+    Map<String, Object> asMap = JsonGenerator.objectsToMapRepresentation(exampleObjects());
+
+    assertEquals(3, asMap.size());
+    assertNotNull(asMap.get("rootPipelineId"));
+    assertNotNull(asMap.get("slots"));
+    assertNotNull(asMap.get("pipelines"));
+  }
+
+  public void testPipelineObjectsToJson() throws Exception {
+
+    //test (fairly pointless, but ensures no loops/etc at least)
+    PipelineObjects example = exampleObjects();
+    String json = stripWhitespace(JsonGenerator.pipelineObjectsToJson(example));
+
+    int length = EXAMPLE_JSON_RESPONSE.length();
```

Comment:
Take a look to this https://www.baeldung.com/jackson-compare-two-json-objects and you can get rid of the whitespace thing... order in the map can be different too

Reply:
problem is also comparison of timestamps and some datastore ids. Seems like in the stubbed LocalDatastore they're using for tests, job ids get set sequentially; eg, when you run this in isolation, the jobId==0, but it can be other things when tests run in batches.